### PR TITLE
Fusion CRM tutorial

### PR DESCRIPTION
**Pull request by marcushellberg**
Here's the Fusion CRM tutorial. It has been reviewed by Tatu and Olli, and gone through Ruth for edits. 

I would like a second pair of eyes on: 
* The actual technical implementation/asciidoc
* Titles. Chapter and tutorial. 

Also, I need help figuring out how to add SEO metadata.

### CODE REVIEW

**Comment by jouni** [comment]
> Also, I need help figuring out how to add SEO metadata.

What exactly do you need? I don’t think we currently have any means of defining those explicitly. Each page uses the first page paragraph as the description and one of the images in the page as the preview image.

**Comment by jouni** [comment]
I’ll try to resolve the conflict, and fix the broken cross references that prevent the build from passing.

File: articles/fusion/tutorials/full-stack/creating-views.adoc [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+   `;
+ }
+}
+----
+
+When building a UI reactively, you define a template using values from the state (`Contact`). Whenever the model changes, the template is automatically updated. 
+
+TIP: As a rule of thumb, you should not use `document.querySelector` or LitElement `@query` to get element reference in order to set values or properties imperatively. There is almost always a way to achieve the same with binding values through the template.
+
+== An introduction to LitElement
```
Comment (web-padawan) [inline articles/fusion/tutorials/full-stack/creating-views.adoc:82]: This part is now covered by "LitElement basics" article, see #245. Do we need to duplicate it also here?

Or is the purpose of this tutorial to cover everything without linking to other docs articles?

File: articles/fusion/tutorials/full-stack/creating-views.adoc [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+ `;
+}
+----
+
+- The text field is configured to show a placeholder text and a clear button.
+- Grid columns map to properties on an object based on the `path` attribute. By default, the header name is derived from the path. You can override it with `header`, for instance when binding to a nested object's path. 
+- Utility CSS classes like `flex` and `h-full` help us to easily create layouts without having to write custom CSS. We will include them in the application in the next step.
+
+== Importing CSS helper classes to the theme
+The Vaadin Lumo theme comes with a https://github.com/vaadin/vaadin-lumo-styles/pull/103[proposed^] set of CSS utility classes to simplify layouting and styling. Instead of having to write your own CSS for common tasks like adding margins, padding, setting sizes, and defining layouts.
```
Comment (web-padawan) [inline articles/fusion/tutorials/full-stack/creating-views.adoc:286]: Do we recommend using a feature that is not officially part of Vaadin (yet) and skip DX tests and PiT for it?

**Review by web-padawan — CHANGES_REQUESTED** [review]

File: articles/fusion/tutorials/full-stack/creating-views.adoc (hunk 1 above)
Comment (jouni) [reply articles/fusion/tutorials/full-stack/creating-views.adoc:82]: Marcus asked about this in Slack. The purpose of a tutorial usually is to be standalone, and not require the reader to bounce back and forth between articles. But it’s just a guideline.

Alternatively, we could try to include the relevant pieces from the LitElement article in here (using `include::[]`)

File: articles/fusion/tutorials/full-stack/creating-views.adoc (hunk 2 above)
Comment (jouni) [reply articles/fusion/tutorials/full-stack/creating-views.adoc:286]: We have a meeting about this today, and have been talking about what to do with them for a while already.

Comment (web-padawan) [reply articles/fusion/tutorials/full-stack/creating-views.adoc:286]: Good to know, thanks.

File: articles/fusion/tutorials/full-stack/creating-views.adoc (hunk 1 above)
Comment (web-padawan) [reply articles/fusion/tutorials/full-stack/creating-views.adoc:82]: Thanks. It would be good to include the relevant pieces indeed.

**Review by web-padawan — COMMENTED** [review]
Resolved my comments to unlock merging. 

But let's try to reuse "LitElement basics" article, if possible.

**Comment by jouni** [comment]
Some observations:

- Vale errors and warnings:
  - All errors should be fixed, for example:
    - All headings should be converted to Title Case (watch out for four letter words like "from" and "with", those are also capitalized)
   - Warnings would be great to address if possible, but not required.
     - All filenames should be written as `[filename]#file.java#`, or at least as `` `file.java` ``
     - Admonitions (notes, warnings, etc.) should have a title, so it’s easier for the reader to decide whether or not it is important for them. And before adding an admonition, consider if it should rather be part of the normal document flow.
- Any links to our components should point to the DS docs, not vaadin.com/componens
- Double check all lists that they format correctly. You need an empty line before and after the list items. Here’s [an example](https://6051b2e5206c1c000787dcbc--vaadin-docs-preview.netlify.app/fusion/tutorials/full-stack/endpoints/#an-introduction-to-spring-boot)
- If you want to make certain images smaller (for example [the one at the end of this article](https://6051b2e5206c1c000787dcbc--vaadin-docs-preview.netlify.app/fusion/tutorials/full-stack/forms-and-validation/#maximizing-the-form-on-narrow-viewports)), use `image::image.png[width=250]` (a suitable pixel width).

**Comment by jouni** [comment]
We can also agree to merge this as is, and continue fixing the Vale errors/warnings from there on.

**Comment by marcushellberg** [comment]
> > Also, I need help figuring out how to add SEO metadata.
> 
> What exactly do you need? I don’t think we currently have any means of defining those explicitly. Each page uses the first page paragraph as the description and one of the images in the page as the preview image.

A way of explicitly defining the meta description would be enough for now. It's not a blocker for going live, but it will affect SEO negatively if the search results in Google don't have a clear summary.

**Comment by jouni** [comment]
> A way of explicitly defining the meta description would be enough for now.

We’ve been planning on adding a way to define the description for an article, as we have other uses for that as well. Haven’t had time to work on it yet, though.

> I was planning on doing it, but I not sure how to do it. 

You need to add special comments, aka tags, around the content that you want to reuse, and then use those tags with with `include::[]` directive. Here are the relevant docs: https://docs.asciidoctor.org/asciidoc/latest/directives/include-tagged-regions/

**Comment by marcushellberg** [comment]
I think that most relevant stuff is passing now. 

I couldn't get the LitElement chapter include to work, so I just linked it. 

I'm still not sure what the tutorial name should be. Full-stack tutorial is pretty meh.

File: articles/fusion/tutorials/_reactive-intro.adoc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+   `;
+ }
+}
+----
+
+When building a UI reactively, you define a template using values from the state (`Contact`). 
+Whenever the model changes, the template is automatically updated. 
+
+[TIP] 
+.Avoid element references in reactive views
```
Comment (jouni) [inline articles/fusion/tutorials/_reactive-intro.adoc:63]:
```suggestion
.Avoid element references in reactive views
[TIP]
```
The Vale check expects the title to be before the block itself.

File: articles/fusion/tutorials/full-stack/creating-components.adoc [hunk 4]
```diff
@@ -82,9 +88,11 @@ Then, add the component to the template right after the `</vaadin-grid>` end tag
 <contact-form class="flex flex-column sb-s"></contact-form>
 ----
 
-Here, we again use the Lumo utility CSS classes to turn the form into a flex container with spacing between components. 
+Here, you again use the Lumo utility CSS classes to turn the form into a flex container with spacing between components. 
+
+As a last step, make the form wider with by adding the following CSS: 
 
-Finally, make the form wider with some CSS: 
+pass:[<!-- vale Vale.Spelling = NO -->]
```
Comment (jouni) [inline articles/fusion/tutorials/full-stack/creating-components.adoc:95]: Why do we need to disable spell checking here? If it’s nagging something about the file name, wrap that in `[filename]#...#` or `` `...` `` to allow Vale to ignore the contents.

Comment (marcushellberg) [reply articles/fusion/tutorials/full-stack/creating-components.adoc:95]: It was nagging about the lowercase css in the filename. I'll update the filename captions to add backticks.

**Pull request merged**